app/src/main/java/app/unbound/android/Themes.kt

```diff
@@ -36,81 +99,119 @@ class Themes : Manager() {
 
-    private fun swizzle(theme: Class<*>, method: String, value: Int) {
-        val implementation = theme.getDeclaredMethod(method)
+    private fun hookSemantic(addon: JsonElement?) {
+        semanticHooks.forEach { it.unhook() } // Remove all previous hooks to let new themes do their thing
 
-        XposedBridge.hookMethod(implementation, object : XC_MethodHook() {
-            override fun beforeHookedMethod(param: MethodHookParam) {
-                param.result = value
+        val getTheme = try {
+            Unbound.info.classLoader
+                .loadClass("com.discord.theme.ThemeManagerKt")
+                .getDeclaredMethod("getTheme")
+        } catch (e: Exception) {
+            Log.e("Unbound", "[Themes] failed to retrieve getTheme(): $e")
+            return
+        }
+
+//        val unwantedMethods = arrayOf("getClass", "getColor", "getColorRes")
+//        for (method in themeClass.methods) {
+//            Log.d("Unbound", "[Themes] Method: ${method.name}, Parameters: ${method.parameterTypes.joinToString()}")
+//            if (method.name.startsWith("get") && method.name !in unwantedMethods) {
+//                semanticSwizzle(themeClass, method.name, Utilities.parseColor("#f0f"), "key")
+//            }
+//        }
+
+        addon?.asJsonObject?.entrySet()?.forEach { (key, json) ->
+            val obj = json.asJsonObject
+
+            val themeClass = getTheme.invoke(null)::class.java
+
+            val segments = key.split("_")
+            val getterMethod = "get" + segments.joinToString("") { it.lowercase().replaceFirstChar(Char::uppercase) }
+
+
+            if (obj.get("type").asString == "color") {
+                val colorValue = obj.get("value").asString
+                val colorOpacity = obj.get("opacity")?.asFloat
+
+                val color = Utilities.parseColor(colorValue, colorOpacity)
+
+                semanticSwizzle(themeClass, getterMethod, color, key)
+
+            } else if (obj.get("type").asString == "raw") {
+                // Unimplemented
+//                val rawKey = obj.get("value").asString
+//                val colorOpacity = obj.get("opacity")?.asFloat
+                return
             }
-        })
+        }
+    }
+    private fun semanticSwizzle(theme: Class<*>, method: String, value: Int?, key: String) {
+        try {
+            if (value != null) {
+                Log.d("Unbound", "[Themes] Applying semantic $key")
+                val implementation = theme.getDeclaredMethod(method)
+                val hook = XposedBridge.hookMethod(implementation, object : XC_MethodHook() {
+                    override fun beforeHookedMethod(param: MethodHookParam) {
+                        param.result = value
+                    }
+                })
+                semanticHooks.add(hook)
+            } else { throw IllegalArgumentException("value parsed to null.") }
+        } catch (e: NoSuchMethodException) { // Common as most semantic strings aren't native

```

Comment:
Instead of catching it's better to prevent. Throwing errors is usually very slow due to stack traces. The predicate above should prevent this from throwing in the first place.